MovingBlocks/DestinationSol
# Initial Integration of NUI (Relates to #455)

# Description
This pull request introduces the NUI UI library to Destination Sol, with the intention of eventually replacing the built-in solution. It also introduces Terasology's console UI into the game and adds a command to display NUI UI screens. This relates to #455.

I have re-created a series of themed widget graphics for Destination Sol, which are roughly based on the existing UI. You can change those or modify them as you see fit, or even replace them entirely. They were designed for testing NUI, as Terasology's UI graphics looked rather out of place in Destination Sol.

Apart from the console screen, none of the pre-existing UI screens have been ported yet. The NUI UI system and the built-in one co-exist, although they may occasionally conflict, as they use a shared `SpriteBatch` instance.
# Testing
##### Configuring NUI
 - Clone my [`nui-extraction-gestalt7`](https://github.com/BenjaminAmos/TeraNUI/tree/nui-extraction-gestalt7) branch and run the command `gradlew publishToMavenLocal` in the cloned project
##### Testing that NUI works
 - Start Destionation Sol and create a new game (any settings should work)
 - Press the ` key to open the console
 - Ensure that the console is visible and shows the correct text
 - Enter some input into the console (e.g. "help")
 - Ensure that a valid response is produced (e.g. a list of commands)
#### Testing NUI's widgets
 - The MigTestScreen from Terasology has been ported as it shows the UI's alignment capabilities, as well as a variety of widgets. You can view it using the command `showNUIScreen engine:migTestScreen`.
 - Ensure that the widgets display correctly (you may have to use Terasology as a reference)
 - Ensure that the widgets can be interacted with correctly (you may have to use Terasology as a reference)
#### Testing that NUI scales
One of the major issues with the existing UI system is that it does not scale very well. UI elements often appear out of alignment when the resolution is changed. The new NUI implementation should not suffer as much from this issue.
 - Follow the instructions from the prior two sections.
 - Close the console and test screen using the `Escape` key.
 - Open the game menu (press `Escape` again) and go back to the main menu.
 - Go into the settings (options) and set change the screen resolution.
 - Continue from your previous game and follow the instructions from the `Testing that NUI works` and `Testing NUI's widgets` sections again. Ensure that the UI appears visually similar and is still functional.
##### Testing with non-discrete resolutions (Windowed mode only)
 - Follow the same instructions as above but instead of changing the resolution in the options menu, open the in-game pause menu and drag the window corners to change the size of the window (or however you would do so natively in your OS).
 - The console UI should still scale correctly, as well as the text within it.
### Testing UI scaling
 - When the game starts, open the `Options->Resolution` menu and press the `NUI UI Scale` button to change the UI scale.
 - Start a game (either create a new one or continue an existing one)
 - Open the console using the ` key
 - The console UI should appear either larger or smaller. The NUI widgets are either being upscaled or downscaled to the target scale factor.
 - Scale factors below 1 should increase the screen size. Scale factors above 1 should decrease it. The screen is scaled from a resolution of `target * scale` to the target resolution.
## Multi-platform testing
 - If you could test these changes on an Android device or emulator, then that would be helpful, as I have not tested these changes on Android extensively. So far as I know though they superficially work though. There may be difficulties using the UI on a touch screen though, as it is fairly desktop-oriented at the moment.
 - I haven't tested this on MacOS or Linux, although they should work with it.
### Reference Renderings
I have taken the following screenshots to show how the NUI UI appears on my set-up. If the appearance differs from these then there might be a rendering bug that I was not aware of.

![ConsoleScreen](https://user-images.githubusercontent.com/24301287/87829602-4b682000-c877-11ea-971f-e6d9fb1808b1.PNG)
![MigTestScreen](https://user-images.githubusercontent.com/24301287/87829609-4c994d00-c877-11ea-99aa-aeb04c051345.PNG)
![AndroidConsoleScreen](https://user-images.githubusercontent.com/24301287/87829951-f11b8f00-c877-11ea-8a27-4a662fa5bb94.png)
![AndroidMigTestScreen](https://user-images.githubusercontent.com/24301287/87829954-f1b42580-c877-11ea-8036-50f1ae372818.png)

# Notes
 - As with Terasology, NUI operates independantly of the entity-system.
 - The only UI screens that have been ported so-far are the console UI, which I moved from Terasology, and the `MigTestScreen`, as it's a very good test of how the widgets render.
 - This relates to MovingBlocks/Terasology#4017 (which removes the shared NUI code from Terasology) but does not depend on it.
 - The extracted NUI code used here is almost entirely the same as in MovingBlocks/TeraNUI#2 but contains a couple of small commits on-top to make it compatible with gestalt 7 (changes [here](https://github.com/BenjaminAmos/TeraNUI/compare/nui-extraction-merge...BenjaminAmos:nui-extraction-gestalt7)). The branch was re-based on-top of the main `nui-extraction-merge` before submitting this pull request.

# Known Issues
## Minor Issues
 - The console UI does not properly scroll at the moment. This is due to the output history buffer being too short.
 - The console UI wraps text too early. I think this is due to the console implementation, rather than NUI itself, as it does not seem affected by UI scale.
 - The UI does not always scale to very small screens. The layout will sometimes be horizontally truncated. This is most prominent when testing the UI under very small resolutions, or on Android.
## Major issues
 - Currently this pull request depends on the use of the `mavenLocal` repository, as the NUI library needs to be built from source for now.
 - The code is not very well documented yet. I will add some JavaDoc to the new classes/methods soon.
 - The NUI reflection work-around in gradle includes `javax.servlet` into the build, which should not be a compile-time dependency, however it was needed to get the code to work.
 - The textures in the `engine/src/main/resources/org/destinationsol/assets/textures/ui/nui/terasology` folder are taken directly from Terasology, when Destination Sol equivalents should be used instead. The UI theme should be consistent, although most of the UI graphics there are unused at the moment.
